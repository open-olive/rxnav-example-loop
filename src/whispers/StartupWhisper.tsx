import { React } from '@oliveai/ldk';
import { stripIndent } from 'common-tags';

// Expect a search handler prop that is the same function used to handle search text from Olive Helps
interface StartupWhisperProps {
  searchHandler: (searchTerm: string) => void;
}
function StartupWhisper({ searchHandler }: StartupWhisperProps) {
  // When creating JSX Whispers, we can use React hooks too!
  const [searchTerm, updateSearchTerm] = React.useState('');

  // Some text for our markdown components to keep the JSX code cleaner
  // stripIndent is a util commonly used to make sure markdown renders as expected
  const headerBody = stripIndent`
    # Look up 194,000+ unique drug names.
    Enter a drug name to find strength, formulation, RXCUI, synonyms, and more (ex. \`Cymbalta\`).
  `;
  const footerBody = stripIndent`
    _This data is powered by the U.S. National Library of Medicine._

    _Please do not search for patient health information as all data searched in this Loop will be sent to the API source._
  `;

  // Return JSX just like in React, but using Whisper-specific components. See: https://docs.oliveai.dev/aptitudes/whisper/jsx-whispers
  return (
    <oh-whisper label="RxNav Drug Lookup" onClose={null}>
      <oh-markdown body={headerBody} />
      <oh-text-input
        label="Search by drug name"
        onChange={(_, val) => {
          // Updating state is identical to standard React
          updateSearchTerm(val);
        }}
      />
      <oh-button
        label="Search"
        onClick={() => {
          // Trigger the search handler with text from the text input field
          searchHandler(searchTerm);
        }}
      />
      <oh-divider />
      <oh-markdown body={footerBody} />
    </oh-whisper>
  );
}

export default StartupWhisper;
