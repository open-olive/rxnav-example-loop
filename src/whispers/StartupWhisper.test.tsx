import { React } from '@oliveai/ldk';
import { create, act } from 'react-test-renderer';

import StartupWhisper from './StartupWhisper';

describe('StartupWhisper', () => {
  it('should render correctly', () => {
    // Use the test renderer to create a tree object that can be snapshot
    const component = create(<StartupWhisper searchHandler={() => null} />);

    // Take advantage of Jest's snapshot feature to ensure the component renders correctly between changes
    // If the component renders differently, the snapshot will fail
    expect(component).toMatchSnapshot();
  });

  it('should call searchHandler with the correct search term', () => {
    const searchHandler = jest.fn();

    // Do some handler triggering to simulate the user typing in the search bar then clicking the button
    const component = create(<StartupWhisper searchHandler={searchHandler} />);
    act(() => component.root.findByType('oh-text-input').props.onChange(undefined, 'foo'));
    act(() => component.root.findByType('oh-button').props.onClick());

    // Make sure our handler is triggered as expected with the updated values coming from user interaction
    expect(searchHandler).toBeCalledWith('foo');
  });
});
