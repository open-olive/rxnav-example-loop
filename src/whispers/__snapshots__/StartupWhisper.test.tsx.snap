// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartupWhisper should render correctly 1`] = `
<oh-whisper
  label="RxNav Drug Lookup"
  onClose={null}
>
  <oh-markdown
    body="# Look up 194,000+ unique drug names.
Enter a drug name to find strength, formulation, RXCUI, synonyms, and more (ex. \`Cymbalta\`)."
  />
  <oh-text-input
    label="Search by drug name"
    onChange={[Function]}
  />
  <oh-button
    label="Search"
    onClick={[Function]}
  />
  <oh-divider />
  <oh-markdown
    body="_This data is powered by the U.S. National Library of Medicine._

_Please do not search for patient health information as all data searched in this Loop will be sent to the API source._"
  />
</oh-whisper>
`;
